feat(split transform): add regex support for separator in split transform

psinghal20 opened the pull request:
Fixes #1881.

The input strings for separator are escaped initially to allow for literal strings using any regex metacharacters as mentioned by @bruceg [here](https://github.com/timberio/vector/issues/1881#issuecomment-649830946).

psinghal20 committed:
feat(split transform): add regex support for separator in split transform

Signed-off-by: psinghal20 <psinghal20@gmail.com>

psinghal20 committed:
chore: run cargo fmt

Signed-off-by: psinghal20 <psinghal20@gmail.com>

bruceg commented on src/transforms/split.rs:
```diff
@@ -82,6 +88,17 @@ impl Split {
             })
             .collect();
 
+        let separator = match separator {
+            Some(s) => match Regex::new(&s) {
+                Ok(expression) => Some(expression),
+                Err(err) => {
+                    error!(message = "Invalid regex expression for separator", %err);
+                    None
+                }
+            },
+            None => None,
+        };
```
This should return an error rather than just displaying it. That way it will prevent the config from being built.

bruceg reviewed (changes requested):
Thanks for this. See my comments for a couple of corrections. I'd also like to see before and after benchmarks to ensure this at least doesn't cause a performance regression.

psinghal20 committed:
refactor(split transform): replace match statement for separator with map and return error in case of invalid regex expression

Signed-off-by: psinghal20 <psinghal20@gmail.com>

psinghal20 committed:
Merge branch 'master' into split-regex

psinghal20 commented:
@bruceg I have made the changes you mentioned. Also, I was trying to run the benchmarks using `cargo bench` but it gets stuck while running the `pipe` benchmark on warming-up stage.

Hoverbear commented:
@psinghal20 You should be able to limit it to a specific benchmark if you need to!

psinghal20 commented:
Hi @bruceg, I ran the benchmarks for the changes. The results can be found [here](https://gist.github.com/psinghal20/311de314d8094bb0b00e4dce817f712f). I think there is a performance hit in some of the benchmarks, though it is not that clear. 

PS: There was some problem while running the normal benchmarks for me, all the benching process were just going into sleep and waiting for some event. Don't know why. I ran the wasm benchmarks for the above results.

Hoverbear commented:
The WASM ones probably won't cover this :(

bruceg commented:
I wasn't clear. There is currently no benchmark to test this. My comment was that we should make a benchmark to show this doesn't cause a performance regression. I'll write something up.

psinghal20 commented:
Oh sorry! I thought you mean the current benchmark numbers. :sweat_smile: 

I will try writing a benchmark similar to the others for split transform. Meanwhile, I wrote a simple benchmark just for the Regex vs String split functions. 
```
Split/Without regex     time:   [64.972 ns 65.680 ns 66.498 ns]
                        change: [+0.3620% +1.9603% +3.5347%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) high mild
  7 (7.00%) high severe
Split/With regex        time:   [90.120 ns 90.973 ns 91.834 ns]
                        change: [-3.9364% -2.4538% -0.9200%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
```
This benchmark doesn't count the regex compilation time and uses the same input string. Considering this, I think it might introduce some performance overhead. I will try to write the benchmark for the split transform tomorrow.

psinghal20 committed:
chore(benchmarks): add benchmark for split transform

Signed-off-by: psinghal20 <psinghal20@gmail.com>

psinghal20 committed:
chore: fix clippy lint errors

Signed-off-by: psinghal20 <psinghal20@gmail.com>

psinghal20 committed:
refactor(benchmark): refactor split transform benchmark similar to wasm benches

Signed-off-by: psinghal20 <psinghal20@gmail.com>

psinghal20 commented:
Hi @bruceg, I wrote a benchmark for the split transform. Here are the results:

Without Regex 
```
split_transform         time:   [939.83 ns 964.64 ns 990.52 ns]
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) high mild
```

With Regex
```
split_transform         time:   [1.4710 us 1.5069 us 1.5491 us]
                        change: [+64.450% +69.399% +74.054%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

You can find the pdf plot for the same here: https://gist.githubusercontent.com/psinghal20/680ae1d0657ae8f11b7381132510a027/raw/34cf4620e882ddf507c1c72f69594dde90c531aa/pdf.svg

I believe this change does introduce performance overhead and we might look into keeping this config literal till we don't require regex explicitly.

@Hoverbear thanks for your help!

psinghal20 committed:
chore: fix clippy warnings

Signed-off-by: psinghal20 <psinghal20@gmail.com>
